File: Xamarin/WritePadSDKAndroidSample/XamarinSDKSample/InkView.cs
/* ************************************************************************************* */
/* *    PhatWare WritePad SDK                                                          * */
/* *    Copyright (c) 2008-2015 PhatWare(r) Corp. All rights reserved.                 * */
/* ************************************************************************************* */

/* ************************************************************************************* *
 *
 * WritePad SDK Xamarin Sample for Android
 *
 * Unauthorized distribution of this code is prohibited. For more information
 * refer to the End User Software License Agreement provided with this
 * software.
 *
 * This source code is distributed and supported by PhatWare Corp.
 * http://www.phatware.com
 *
 * THIS SAMPLE CODE CAN BE USED  AS A REFERENCE AND, IN ITS BINARY FORM,
 * IN THE USER'S PROJECT WHICH IS INTEGRATED WITH THE WRITEPAD SDK.
 * ANY OTHER USE OF THIS CODE IS PROHIBITED.
 *
 * THE MATERIAL EMBODIED ON THIS SOFTWARE IS PROVIDED TO YOU "AS-IS"
 * AND WITHOUT WARRANTY OF ANY KIND, EXPRESS, IMPLIED OR OTHERWISE,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTY OF MERCHANTABILITY OR
 * FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT SHALL PHATWARE CORP.
 * BE LIABLE TO YOU OR ANYONE ELSE FOR ANY DIRECT, SPECIAL, INCIDENTAL,
 * INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND, OR ANY DAMAGES WHATSOEVER,
 * INCLUDING WITHOUT LIMITATION, LOSS OF PROFIT, LOSS OF USE, SAVINGS
 * OR REVENUE, OR THE CLAIMS OF THIRD PARTIES, WHETHER OR NOT PHATWARE CORP.
 * HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH LOSS, HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE
 * POSSESSION, USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * US Government Users Restricted Rights
 * Use, duplication, or disclosure by the Government is subject to
 * restrictions set forth in EULA and in FAR 52.227.19(c)(2) or subparagraph
 * (c)(1)(ii) of the Rights in Technical Data and Computer Software
 * clause at DFARS 252.227-7013 and/or in similar or successor
 * clauses in the FAR or the DOD or NASA FAR Supplement.
 * Unpublished-- rights reserved under the copyright laws of the
 * United States.  Contractor/manufacturer is PhatWare Corp.
 * 1314 S. Grand Blvd. Ste. 2-175 Spokane, WA 99202
 *
 * ************************************************************************************* */

using System;
using System.Collections.Generic;
using System.Linq;
using Android.Content;
using Android.Graphics;
using Android.Runtime;
using Android.Util;
using Android.Views;
using Android.Widget;

namespace WritePadXamarinSample
{
    class InkView: View
    {
        private Path mPath;
        private int mCurrStroke;
        private Paint mPaint;
        private Paint mResultPaint;
		private Path gridpath;

        private LinkedList<Path> mPathList;
        private float mX, mY;
        private const float TOUCH_TOLERANCE = 2;
        private bool mMoved;
		private List<WritePadAPI.CGTracePoint> currentStroke;
		private int strokeLen = 0;
		struct FPoint
		{
			public float X;
			public float Y; 
		};
		private FPoint _lastPoint;
		private FPoint _previousLocation;

		private const int SEGMENT2 = 2;
		private const int SEGMENT3 = 3;

		private const float GRID_GAP = 65;
		private const int SEGMENT4 = 4;

		private const float  SEGMENT_DIST_1   = 3;
		private const float  SEGMENT_DIST_2   = 6;
		private const float  SEGMENT_DIST_3   = 12;
        private int AddPixelsXY( float X, float Y, bool bLastPoint )
		// this method called from inkCollectorThread
		{
			float	xNew, yNew, x1, y1;
			int		nSeg = SEGMENT3;

			if ( mCurrStroke < 0 )
				return 0;

			if  ( strokeLen < 1 )  
			{
				_lastPoint.X = _previousLocation.X = X;
				_lastPoint.Y = _previousLocation.Y = Y;
                WritePadAPI.recoAddPixel(mCurrStroke, X, Y);
                AddCurrentPoint(X, Y);
                strokeLen = 1;
				return  1;
			}

			float dx = Math.Abs( X - _lastPoint.X );
			float dy = Math.Abs( Y - _lastPoint.Y );
			if  ( (dx + dy) < SEGMENT_DIST_1 )  
			{
				_lastPoint.X = _previousLocation.X = X;
				_lastPoint.Y = _previousLocation.Y  = Y;
                WritePadAPI.recoAddPixel(mCurrStroke, X, Y);
                AddCurrentPoint(X, Y);
                strokeLen++;
				return  1;
			}

			if ( (dx + dy) < SEGMENT_DIST_2 )  
				nSeg = SEGMENT2;
			else if ( (dx + dy) < SEGMENT_DIST_3 )
				nSeg = SEGMENT3;
			else
				nSeg = SEGMENT4;
			int     nPoints = 0;
			for ( int i = 1; i < nSeg;  i++ )  
			{
				x1 = _previousLocation.X + ((X - _previousLocation.X)*i ) / nSeg;  //the point "to look at"
				y1 = _previousLocation.Y + ((Y - _previousLocation.Y)*i ) / nSeg;  //the point "to look at"

				xNew = _lastPoint.X + (x1 - _lastPoint.X) / nSeg;
				yNew = _lastPoint.Y + (y1 - _lastPoint.Y) / nSeg;

				if ( xNew != _lastPoint.X || yNew != _lastPoint.Y )
				{
					_lastPoint.X = xNew;
					_lastPoint.Y = yNew;
                    WritePadAPI.recoAddPixel(mCurrStroke, xNew, yNew);
                    AddCurrentPoint(X, Y);
                    strokeLen++;
					nPoints++;
				}
			}

			if ( bLastPoint )  
			{
				// add last point
				if ( X != _lastPoint.X || Y != _lastPoint.Y )  
				{
					_lastPoint.X = X;
					_lastPoint.Y = Y;
                    WritePadAPI.recoAddPixel(mCurrStroke, X, Y);
                    AddCurrentPoint(X, Y);
                    strokeLen++;
					nPoints++;
				}
			}

			_previousLocation.X = X;
			_previousLocation.Y = Y;
			return nPoints;
		}


        protected override void OnDraw(Canvas canvas) 
        {
			// draw grid lines
			mPaint.Color = new Color(255, 0, 0);
			mPaint.StrokeWidth = 1;

			for ( float y = GRID_GAP; y < canvas.Height; y += GRID_GAP) 
			{
				gridpath.Reset ();
				gridpath.MoveTo (0, y);
				gridpath.LineTo (canvas.Width, y);
				canvas.DrawPath(gridpath, mPaint);
			}
			mPaint.Color = new Color(0, 0, 255);
			mPaint.StrokeWidth = 3;

			// draw strokes
            foreach(var aMPathList in mPathList) 
            {
                canvas.DrawPath(aMPathList, mPaint);
            }
            canvas.DrawPath(mPath, mPaint);
        }
        private void AddCurrentPoint(float mX, float mY)
        {
            var point = new WritePadAPI.CGTracePoint();
            point.pressure = WritePadAPI.DEFAULT_INK_PRESSURE;
            point.pt = new WritePadAPI.CGPoint();
            point.pt.x = mX;
            point.pt.y = mY;
            currentStroke.Add(point);
        }

        private void touch_start(float x, float y)
        {
            mPath.Reset();
            currentStroke = new List<WritePadAPI.CGTracePoint>();
            mPath.MoveTo(x, y);
            mX = x;
            mY = y;
            AddCurrentPoint(mX, mY);
            mMoved = false;
            mCurrStroke = WritePadAPI.recoNewStroke(3, 0xFFFF0000);
			strokeLen = 0;
			AddPixelsXY(mX, mY, false);
        }

        private void touch_move(float x, float y)
        {
            float dx = Math.Abs(x - mX);
            float dy = Math.Abs(y - mY);
            if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE)
            {
                mPath.QuadTo(mX, mY, (x + mX) / 2, (y + mY) / 2);
                mMoved = true;
                mX = x;
                mY = y;
                AddCurrentPoint(mX, mY);			
				AddPixelsXY(mX, mY, false);
            }
        }

		private void touch_up(float x, float y)
		{
			var gesture = WritePadAPI.detectGesture(WritePadAPI.GEST_RETURN | WritePadAPI.GEST_CUT | WritePadAPI.GEST_BACK, currentStroke);
            
			AddPixelsXY(x, y, true);
            AddCurrentPoint(mX, mY);			
			mCurrStroke = -1;
			mMoved = false;

			if (!mMoved)
				mX++;
			mPath.LineTo(mX, mY);
			mPathList.AddLast(mPath);
			mPath = new Path();
			Invalidate();

			switch (gesture)
            {
                case WritePadAPI.GEST_RETURN:
                    if (OnReturnGesture != null)
                    {
                        mPathList.RemoveLast();
                        WritePadAPI.recoDeleteLastStroke();
                        Invalidate(); 
                        OnReturnGesture();  
						return;
                    }
                    break;
                case WritePadAPI.GEST_CUT:
                    if (OnCutGesture != null)
                    {
                        mPathList.RemoveLast();
                        WritePadAPI.recoDeleteLastStroke();
                        Invalidate(); 
                        OnCutGesture(); 
						return;
                    }
                    break;

				case WritePadAPI.GEST_BACK_LONG :
					mPathList.RemoveLast();
					WritePadAPI.recoDeleteLastStroke();
					if ( WritePadAPI.recoStrokeCount() > 0 )
					{
						mPathList.RemoveLast();
						WritePadAPI.recoDeleteLastStroke();
					}
					Invalidate(); 
					return;
            }

		}

		public struct WordAlternative
        {
            public string Word;
            public int Weight;
        }
		public string Recognize( bool bLearn )
        {
            var res = "";
            var resultStringList = new List<string>();
			var wordList = new List<List<WordAlternative>>();
            var count = WritePadAPI.recoStrokeCount();
            string defaultResult = WritePadAPI.recoInkData(count, false, false, false, false);
			// can also use the default result
            resultStringList.Add(defaultResult);
            var wordCount = WritePadAPI.recoResultColumnCount();
            for (var i = 0; i < wordCount; i++)
            {
				var wordAlternativesList = new List<WordAlternative>();
                var altCount = WritePadAPI.recoResultRowCount(i);
                for (var j = 0; j < altCount; j++)
                {
                    var word = WritePadAPI.recoResultWord(i, j);
                    if (word == "<--->")
                        word = "*Error*";
                    if (string.IsNullOrEmpty(word))
                        continue;
					var weight = WritePadAPI.recoResultWeight(i, j);
					var flags = WritePadAPI.recoGetFlags();
				
					if ( j == 0 && bLearn && weight > 75 && 0 != (flags & WritePadAPI.FLAG_ANALYZER) )
					{
						WritePadAPI.recoLearnWord( word, weight );
					}
                    if (wordAlternativesList.All(x => x.Word != word))
                    {
                        wordAlternativesList.Add(new WordAlternative
                        {
                            Word = word,
                            Weight = weight
                        });
					}
					while (resultStringList.Count < j + 2)
                            resultStringList.Add("");
						if (resultStringList[j+1].Length > 0)
							resultStringList[j+1] += "\t\t";
						resultStringList[j+1] += word + "\t[" + weight + "%]";
                    
                }
				wordList.Add(wordAlternativesList);
            }
            foreach (var line in resultStringList)
            {
                if(string.IsNullOrEmpty(line))
                    continue;                
                if (res.Length > 0)
                {
                    res += Environment.NewLine;
                }
                res += line;                
            }
            return res;
        }
			
        public void cleanView(bool emptyAll)
        {
            WritePadAPI.recoResetInk();
            mCurrStroke = -1;
            mPathList.Clear();
            mPath.Reset();            
            Invalidate();
        }

        public override bool OnTouchEvent(MotionEvent ev) 
        {
            float x = ev.GetX();
            float y = ev.GetY();

            switch (ev.Action) 
            {
                case MotionEventActions.Down:
                    touch_start(x, y);
                    Invalidate();
                    break;

                case MotionEventActions.Move:
                    for (int i = 0, n = ev.HistorySize; i < n; i++) 
					{
						touch_move(ev.GetHistoricalX(i), ev.GetHistoricalY(i));
                    }
                    touch_move(x, y);
                    Invalidate();
                    break;

                case MotionEventActions.Up:
					touch_up(x, y);
                    Invalidate();
                    break;
            }
            return true;
        }

        protected InkView(IntPtr javaReference, JniHandleOwnership transfer) : base(javaReference, transfer)
        {
           
        }

        public InkView(Context context) : base(context)
        {
            
        }

        public InkView(Context context, IAttributeSet attrs) : base(context, attrs)
        {
            mPath = new Path();
            mPathList = new LinkedList<Path>();
            mCurrStroke = -1;
            mPaint = new Paint();
            mPaint.AntiAlias = true;
            mPaint.Dither = true;
            mPaint.Color = new Color(0, 0, 255);
            mPaint.SetStyle(Paint.Style.Stroke);
            mPaint.StrokeJoin = Paint.Join.Round;
            mPaint.StrokeCap = Paint.Cap.Round;
            mPaint.StrokeWidth = 3;

            mResultPaint = new Paint();
            mResultPaint.TextSize = 32;
            mResultPaint.AntiAlias = true;
            mResultPaint.SetARGB(0xff, 0x00, 0x00, 0x00);

			gridpath = new Path();
        }

        public InkView(Context context, IAttributeSet attrs, int defStyle) : base(context, attrs, defStyle)
        {
            
        }

        public event Action OnReturnGesture;
        public event Action OnCutGesture;
    }
}

File: Android/sample_astudio/app/src/main/java/com/phatware/android/recotest/InkView.java
/* ************************************************************************************* */
/* *    PhatWare WritePad SDK                                                          * */
/* *    Copyright (c) 1997-2015 PhatWare(r) Corp. All rights reserved.                 * */
/* ************************************************************************************* */

/* ************************************************************************************* *
 *
 * WritePad Android Sample
 *
 * Unauthorized distribution of this code is prohibited. For more information
 * refer to the End User Software License Agreement provided with this 
 * software.
 *
 * This source code is distributed and supported by PhatWare Corp.
 * http://www.phatware.com
 *
 * THIS SAMPLE CODE CAN BE USED  AS A REFERENCE AND, IN ITS BINARY FORM, 
 * IN THE USER'S PROJECT WHICH IS INTEGRATED WITH THE WRITEPAD SDK. 
 * ANY OTHER USE OF THIS CODE IS PROHIBITED.
 * 
 * THE MATERIAL EMBODIED ON THIS SOFTWARE IS PROVIDED TO YOU "AS-IS"
 * AND WITHOUT WARRANTY OF ANY KIND, EXPRESS, IMPLIED OR OTHERWISE,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTY OF MERCHANTABILITY OR
 * FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT SHALL PHATWARE CORP.  
 * BE LIABLE TO YOU OR ANYONE ELSE FOR ANY DIRECT, SPECIAL, INCIDENTAL, 
 * INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND, OR ANY DAMAGES WHATSOEVER, 
 * INCLUDING WITHOUT LIMITATION, LOSS OF PROFIT, LOSS OF USE, SAVINGS 
 * OR REVENUE, OR THE CLAIMS OF THIRD PARTIES, WHETHER OR NOT PHATWARE CORP.
 * HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH LOSS, HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE
 * POSSESSION, USE OR PERFORMANCE OF THIS SOFTWARE.
 * 
 * US Government Users Restricted Rights 
 * Use, duplication, or disclosure by the Government is subject to
 * restrictions set forth in EULA and in FAR 52.227.19(c)(2) or subparagraph
 * (c)(1)(ii) of the Rights in Technical Data and Computer Software
 * clause at DFARS 252.227-7013 and/or in similar or successor
 * clauses in the FAR or the DOD or NASA FAR Supplement.
 * Unpublished-- rights reserved under the copyright laws of the
 * United States.  Contractor/manufacturer is PhatWare Corp.
 * 10414 W. Highway 2, Ste 4-121 Spokane, WA 99224
 *
 * ************************************************************************************* */

package com.phatware.android.recotest;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import com.phatware.android.DialogHelper;
import com.phatware.android.RecoInterface.WritePadAPI;
import com.phatware.android.WritePadManager;

import java.util.LinkedList;

public class InkView extends View implements RecoTest.OnInkViewListener {
    private Path mPath;
    private int mCurrStroke;
    private Paint mPaint;
    private Paint mResultPaint;
    private LinearLayout recognizedTextContainer;
    private TextView readyText;
    private LinkedList<Path> mPathList;
    private float mX, mY;
    private static final float TOUCH_TOLERANCE = 2;
    private boolean mMoved;
    private String lastResult = null;
    
	private Path  gridpath = new Path();
    
    private final float GRID_GAP = 65;

    // Define the Handler that receives messages from the thread and update the progress
    private final Handler mHandler = new Handler() {
        public void handleMessage(Message msg) {
            recognizedTextContainer.removeAllViews();
            Bundle b = msg.getData();
            lastResult = b.getString( "result" );
            int words = WritePadManager.recoResultColumnCount();
            for (int w = 0; w < words; w++) {
                int alternatives = WritePadManager.recoResultRowCount(w);
                if (alternatives > 0) {

                    final CharSequence[] alternativesCollection = new CharSequence[alternatives];
                    for (int a = 0; a < alternatives; a++) {
                        String word = WritePadManager.recoResultWord(w, a);
                        alternativesCollection[a] = word;
                    }

                    Button word = new Button(getContext());
                    word.setTransformationMethod(null);
                    word.setText(WritePadManager.recoResultWord(w, 0));
                    word.setOnClickListener(new OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            if (alternativesCollection.length > 0)
                                DialogHelper.createAlternativesDialog(getContext(), getWindowToken(), alternativesCollection).show();
                        }
                    });
                    recognizedTextContainer.addView(word);

                }
            }
        }
    };

    public void cleanView(boolean emptyAll) {
        WritePadManager.recoResetInk();
        mCurrStroke = -1;
        mPathList.clear();
        mPath.reset();
        recognizedTextContainer.removeAllViews();
        if (emptyAll) {
            readyText.setText("");
        }
        invalidate();
    }

    public Handler getHandler() {
        return mHandler;
    }

    public InkView(Context context) {
        this(context, null, 0);
    }

    public InkView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public InkView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        mPath = new Path();
        mPathList = new LinkedList<Path>();
        mCurrStroke = -1;
        mPaint = new Paint();
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setDither(true);
        mPaint.setColor(0xFF0000FF);
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setStrokeJoin(Paint.Join.ROUND);
        mPaint.setStrokeCap(Paint.Cap.ROUND);
        mPaint.setStrokeWidth(3);

        mResultPaint = new Paint();
        mResultPaint.setTextSize(32);
        mResultPaint.setAntiAlias(true);
        mResultPaint.setARGB(0xff, 0x00, 0x00, 0x00);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        mPaint.setColor(0xFFFF0000);
        mPaint.setStrokeWidth(1);

        for ( float y = GRID_GAP; y < canvas.getHeight(); y += GRID_GAP) 
		{
			gridpath.reset ();
			gridpath.moveTo (0, y);
			gridpath.lineTo (canvas.getWidth(), y);
			canvas.drawPath(gridpath, mPaint);
		}
        mPaint.setColor(0xFF0000FF);
        mPaint.setStrokeWidth(3);
    	
    	
        for (Path aMPathList : mPathList) {
            canvas.drawPath(aMPathList, mPaint);
        }
        canvas.drawPath(mPath, mPaint);
    }

    private void touch_start(float x, float y) {
        mPath.reset();
        mPath.moveTo(x, y);
        mX = x;
        mY = y;
        mMoved = false;
        mCurrStroke = WritePadManager.recoNewStroke(3.0f, 0xFF0000FF);
        if (mCurrStroke >= 0) {
            int res = WritePadManager.recoAddPixel(mCurrStroke, x, y);
            if (res < 1) {
                // TODO: error
            }
        }
    }
    

    private void touch_move(float x, float y) {
        float dx = Math.abs(x - mX);
        float dy = Math.abs(y - mY);
        if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE) {
            mPath.quadTo(mX, mY, (x + mX) / 2, (y + mY) / 2);
            mMoved = true;
            mX = x;
            mY = y;
            if (mCurrStroke >= 0) {
                int res = WritePadManager.recoAddPixel(mCurrStroke, x, y);
                if (res < 1) {
                    // TODO: error
                }
            }
        }
    }

    private void touch_up() {
        // stopRecognizer();

        mCurrStroke = -1;
        if (!mMoved)
            mX++;
        mMoved = false;
        mPath.lineTo(mX, mY);
        mPathList.add(mPath);
        mPath = new Path();
        invalidate();

        RecoTest rt = (RecoTest) getContext();
        int nStrokeCnt = WritePadManager.recoStrokeCount();
        if (nStrokeCnt == 1) {
            int gesturetype = WritePadAPI.GEST_DELETE + WritePadAPI.GEST_RETURN + WritePadAPI.GEST_SPACE +
                    WritePadAPI.GEST_TAB + WritePadAPI.GEST_BACK + WritePadAPI.GEST_UNDO;
            gesturetype = WritePadManager.recoGesture(gesturetype);
            if (gesturetype != WritePadAPI.GEST_NONE) {
                // TODO: process gesture
                WritePadManager.recoDeleteLastStroke();
                mPathList.removeLast();
                return;
            }
        } else if (nStrokeCnt > 1) {
            int gesturetype = WritePadAPI.GEST_CUT + WritePadAPI.GEST_BACK + WritePadAPI.GEST_RETURN;
            gesturetype = WritePadManager.recoGesture(gesturetype);
            if (gesturetype != WritePadAPI.GEST_NONE && gesturetype != WritePadAPI.GEST_BACK) {
                // TODO: process gesture
                WritePadManager.recoDeleteLastStroke();
                mPathList.removeLast();
                switch (gesturetype) {
                    // case WritePadAPI.GEST_BACK:
                    case WritePadAPI.GEST_BACK_LONG:
                        WritePadManager.recoDeleteLastStroke();
                        mPathList.removeLast();
                        if (WritePadManager.recoStrokeCount() < 1) {
                            recognizedTextContainer.removeAllViews();
                        }

                        rt.mBoundService.dataNotify(WritePadManager.recoStrokeCount());
                        return;

                    case WritePadAPI.GEST_CUT:
                        cleanView(false);
                        lastResult = null;
                        return;

                    case WritePadAPI.GEST_RETURN:
                        sendText();
                        cleanView(false);
                        lastResult = null;
                        return;
                        
                    default :
                    	break;
                }
            }
        }

        // notify recognizer thread about data availability
        rt.mBoundService.dataNotify(nStrokeCnt);
    }

    private void sendText() {
    	if ( lastResult == null )
    	{
	        int count = recognizedTextContainer.getChildCount();
	        StringBuilder stringBuilder = new StringBuilder();
	        for (int i = 0; i < count; i++) {
	            Button button = (Button) recognizedTextContainer.getChildAt(i);
	            stringBuilder.append(button.getText());
	            stringBuilder.append(" ");	
	        }
	        readyText.setText(String.format("%s%s", readyText.getText(), stringBuilder.toString()));
    	}
    	else
    	{
	        readyText.setText(String.format("%s %s", readyText.getText(), lastResult ));    		
    	}
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float x = event.getX();
        float y = event.getY();

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                touch_start(x, y);
                invalidate();
                break;

            case MotionEvent.ACTION_MOVE:
                for (int i = 0, n = event.getHistorySize(); i < n; i++) {
                    touch_move(event.getHistoricalX(i),
                            event.getHistoricalY(i));
                }
                touch_move(x, y);
                invalidate();
                break;

            case MotionEvent.ACTION_UP:
                touch_up();
                invalidate();
                break;
        }
        return true;
    }

    public void setRecognizedTextContainer(LinearLayout recognizedTextContainer) {
        this.recognizedTextContainer = recognizedTextContainer;
    }

    public void setReadyText(TextView readyText) {
        this.readyText = readyText;
    }
}
